Python/immutability.c

```diff
@@ -0,0 +1,376 @@
+
+#include "Python.h"
+#include <ctype.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include "pycore_dict.h"
+#include "pycore_object.h"
+#include "pycore_immutability.h"
+
+
+static PyObject* pop(PyObject* s){
+    Py_ssize_t size = PyList_Size(s);
+    if(size == 0){
+        return NULL;
+    }
+
+    PyObject* item = PyList_GetItem(s, size - 1);
+    if(item == NULL){
+        return NULL;
+    }
+
+    if(PyList_SetSlice(s, size - 1, size, NULL)){
+        return NULL;
+    }
+
+    Py_INCREF(item);
+    return item;
+}
+
+static bool is_c_wrapper(PyObject* obj){
+    return PyCFunction_Check(obj) || Py_IS_TYPE(obj, &_PyMethodWrapper_Type) || Py_IS_TYPE(obj, &PyWrapperDescr_Type);
+}
+
+#define _Py_VISIT_FUNC_ATTR(attr, frontier) do { \
+    if(attr != NULL && !_Py_IsImmutable(attr)){ \
+        if(PyList_Append(frontier, attr)){ \
+            return PyErr_NoMemory(); \
+        } \
+    } \
+} while(0)
+
+
+/**
+ * Special function for walking the reachable graph of a function object.
+ *
+ * This is necessary because the function object has a pointer to the global
+ * object, and this is problematic because freezing any function will make the
+ * global object immutable, which is not always the desired behaviour.
```

Comment:
Is this comment still up to date with the semantics for function freezing we settled on?

This is a small suggestion for reformulation if we keep the current comment.

```suggestion
 * This is necessary because the function object has a pointer to the global
 * object. Freezing it like any other object would also make the
 * global object immutable, which is usually the undesired behaviour.
```

Reply:
yup out of date comment